src/MessageSystem.cpp

```diff
@@ -0,0 +1,346 @@
+/*
+ *  Copyright (C) 2018 SPMod Development Team
+ *
+ *  This file is part of SPMod.
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "spmod.hpp"
+
+void Message::init(int dest, int type, const float *origin, edict_t *edict)
+{
+    m_dest = dest;
+    m_type = type;
+    m_origin = Vector(origin[0], origin[1], origin[2]);
+    m_edict = std::make_unique<Edict>(edict);
+}
+
+void Message::exec() const
+{
+    MESSAGE_BEGIN(m_dest, m_type, m_origin, m_edict->getInternalEdict());
+
+    for (auto param : m_params)
+    {
+        switch (param.type)
```

Comment:
This looks awkward, can't `type` be combined with variants of `data`? So we can just use `std::visit`.